pire/half_final_fsm.h

```diff
@@ -2,5 +2,47 @@
 #include "defs.h"
 
 namespace Pire {
-	Fsm CreateHalfFinalFsm(Fsm fsm, bool addFinalTransitions = false, bool count = false);
+	class HalfFinalFsm {
+	public:
+		HalfFinalFsm(const Fsm& sourceFsm) : fsm(sourceFsm) {}
+
+		void MakeScanner();
+
+		/// Non greedy counter without allowed intersects works correctly on all regexps
+		/// Non simplified non greedy counter with allowed intersects counts number of positions in string,
+		/// on which ends at least one substring that matches regexp
+		/// Simplified non greedy counter with allowed intersects does not always work correctly,
+		/// but has fewer number of states and more regexps can be glued into single scanner
+		void MakeNonGreedyCounter(bool allowIntersects = true, bool simplify = true);
```

Comment:
IMHO
I would prefer if greediness would be a characteristic of a particular quantifier.
(Like usually in regexps, eg. http://www.rexegg.com/regex-quantifiers.html#lazy .)
But I suggest to leave it as it is for now.